File: scheduler.py
import argparse
import random
from dataclasses import dataclass, field
from typing import Dict, List

DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
SHIFTS = ["morning", "afternoon", "evening"]
MIN_PER_SHIFT = 2
MAX_DAYS_PER_EMPLOYEE = 5
CAPACITY_PER_SHIFT = 2


@dataclass
class Employee:
    name: str
    preferences: Dict[str, List[str]]
    assigned_days: int = 0
    assignments: Dict[str, str] = field(default_factory=dict)


def collect_preferences() -> List[Employee]:
    employees: List[Employee] = []
    print("Enter employee names and shift preferences.")
    print("Leave the name blank to finish entering employees.")

    while True:
        name = input("Employee name: ").strip()
        if not name:
            break

        preferences: Dict[str, List[str]] = {}
        for day in DAYS:
            print(f"  {day} preferences:")
            day_pref: List[str] = []
            for rank in range(1, len(SHIFTS) + 1):
                while True:
                    pref = input(f"    Rank {rank} shift (morning/afternoon/evening): ").strip().lower()
                    if pref in SHIFTS and pref not in day_pref:
                        day_pref.append(pref)
                        break
                    print("    Invalid or duplicate shift. Try again.")
            preferences[day] = day_pref

        employees.append(Employee(name=name, preferences=preferences))
        print("")

    return employees


def demo_employees() -> List[Employee]:
    return [
        Employee("Alice", {day: ["morning", "afternoon", "evening"] for day in DAYS}),
        Employee("Bob", {day: ["afternoon", "evening", "morning"] for day in DAYS}),
        Employee("Carol", {day: ["evening", "morning", "afternoon"] for day in DAYS}),
        Employee("Dan", {day: ["morning", "evening", "afternoon"] for day in DAYS}),
        Employee("Ella", {day: ["afternoon", "morning", "evening"] for day in DAYS}),
        Employee("Frank", {day: ["morning", "afternoon", "evening"] for day in DAYS}),
        Employee("Gina", {day: ["evening", "afternoon", "morning"] for day in DAYS}),
        Employee("Hugo", {day: ["afternoon", "morning", "evening"] for day in DAYS}),
        Employee("Ivy", {day: ["morning", "evening", "afternoon"] for day in DAYS}),
    ]


def initialize_schedule() -> Dict[str, Dict[str, List[str]]]:
    return {day: {shift: [] for shift in SHIFTS} for day in DAYS}


def can_work(emp: Employee, day: str) -> bool:
    return emp.assigned_days < MAX_DAYS_PER_EMPLOYEE and day not in emp.assignments


def assign_shift(emp: Employee, day: str, shift: str, schedule: Dict[str, Dict[str, List[str]]]) -> None:
    schedule[day][shift].append(emp.name)
    emp.assignments[day] = shift
    emp.assigned_days += 1


def schedule_day(day: str, employees: List[Employee], schedule: Dict[str, Dict[str, List[str]]]) -> None:
    available = [emp for emp in employees if can_work(emp, day)]
    available.sort(key=lambda emp: (emp.assigned_days, emp.name))

    for emp in available:
        for pref in emp.preferences[day]:
            if len(schedule[day][pref]) < CAPACITY_PER_SHIFT:
                assign_shift(emp, day, pref, schedule)
                break

    for shift in SHIFTS:
        while len(schedule[day][shift]) < MIN_PER_SHIFT:
            candidates = [emp for emp in employees if can_work(emp, day) and emp.name not in sum((schedule[day][s] for s in SHIFTS), [])]
            if not candidates:
                break
            candidates.sort(key=lambda emp: (emp.preferences[day].index(shift) if shift in emp.preferences[day] else len(SHIFTS), emp.assigned_days, emp.name))
            assign_shift(candidates[0], day, shift, schedule)


def schedule_week(employees: List[Employee]) -> Dict[str, Dict[str, List[str]]]:
    schedule = initialize_schedule()
    for day in DAYS:
        schedule_day(day, employees, schedule)
    return schedule


def print_schedule(schedule: Dict[str, Dict[str, List[str]]]) -> None:
    print("\nFinal Weekly Schedule:\n")
    for day in DAYS:
        print(f"{day}:")
        for shift in SHIFTS:
            names = schedule[day][shift]
            print(f"  {shift.capitalize()}: {', '.join(names) if names else '(no assignment)'}")
        print("")


def print_employee_summary(employees: List[Employee]) -> None:
    print("Employee assignment summary:")
    for emp in sorted(employees, key=lambda e: e.name):
        assigned_days = len(emp.assignments)
        print(f"  {emp.name}: {assigned_days} days")
    print("")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Employee shift scheduler")
    parser.add_argument("--demo", action="store_true", help="Run the scheduler with sample employees")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    employees = demo_employees() if args.demo else collect_preferences()
    if not employees:
        print("No employees entered. Exiting.")
        return

    schedule = schedule_week(employees)
    print_schedule(schedule)
    print_employee_summary(employees)


if __name__ == "__main__":
    main()

File: scheduler.cpp
#include <algorithm>
#include <iostream>
#include <map>
#include <random>
#include <string>
#include <vector>

using namespace std;

const vector<string> DAYS = {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"};
const vector<string> SHIFTS = {"morning", "afternoon", "evening"};
const int MIN_PER_SHIFT = 2;
const int MAX_DAYS_PER_EMPLOYEE = 5;
const int CAPACITY_PER_SHIFT = 2;

struct Employee {
    string name;
    map<string, vector<string>> preferences;
    map<string, string> assignments;
    int assigned_days = 0;
};

using Schedule = map<string, map<string, vector<string>>>;

string normalize_shift(string value) {
    transform(value.begin(), value.end(), value.begin(), ::tolower);
    return value;
}

bool valid_shift(const string& shift) {
    return find(SHIFTS.begin(), SHIFTS.end(), shift) != SHIFTS.end();
}

vector<Employee> collect_preferences() {
    vector<Employee> employees;
    cout << "Enter employee names and ranked shift preferences. Leave name blank to use sample data." << endl;
    while (true) {
        cout << "Employee name: ";
        string name;
        getline(cin, name);
        if (name.empty()) break;

        Employee emp;
        emp.name = name;
        for (const string& day : DAYS) {
            cout << "  " << day << " preferences:" << endl;
            vector<string> day_pref;
            while (day_pref.size() < SHIFTS.size()) {
                cout << "    Rank " << day_pref.size() + 1 << " shift (morning/afternoon/evening): ";
                string pref;
                getline(cin, pref);
                pref = normalize_shift(pref);
                if (valid_shift(pref) && find(day_pref.begin(), day_pref.end(), pref) == day_pref.end()) {
                    day_pref.push_back(pref);
                } else {
                    cout << "    Invalid or duplicate shift. Try again." << endl;
                }
            }
            emp.preferences[day] = day_pref;
        }
        employees.push_back(emp);
        cout << endl;
    }
    return employees;
}

vector<Employee> demo_employees() {
    auto make_prefs = [&](const string& first, const string& second, const string& third) {
        map<string, vector<string>> prefs;
        for (const auto& day : DAYS) {
            prefs[day] = {first, second, third};
        }
        return prefs;
    };

    vector<Employee> employees;
    employees.push_back({"Alice", make_prefs("morning", "afternoon", "evening"), {}, 0});
    employees.push_back({"Bob", make_prefs("afternoon", "evening", "morning"), {}, 0});
    employees.push_back({"Carol", make_prefs("evening", "morning", "afternoon"), {}, 0});
    employees.push_back({"Dan", make_prefs("morning", "evening", "afternoon"), {}, 0});
    employees.push_back({"Ella", make_prefs("afternoon", "morning", "evening"), {}, 0});
    employees.push_back({"Frank", make_prefs("morning", "afternoon", "evening"), {}, 0});
    employees.push_back({"Gina", make_prefs("evening", "afternoon", "morning"), {}, 0});
    employees.push_back({"Hugo", make_prefs("afternoon", "morning", "evening"), {}, 0});
    employees.push_back({"Ivy", make_prefs("morning", "evening", "afternoon"), {}, 0});
    return employees;
}

Schedule initialize_schedule() {
    Schedule schedule;
    for (const string& day : DAYS) {
        for (const string& shift : SHIFTS) {
            schedule[day][shift] = {};
        }
    }
    return schedule;
}

bool can_work(const Employee& emp, const string& day) {
    return emp.assigned_days < MAX_DAYS_PER_EMPLOYEE && emp.assignments.find(day) == emp.assignments.end();
}

void assign_shift(Employee& emp, const string& day, const string& shift, Schedule& schedule) {
    schedule[day][shift].push_back(emp.name);
    emp.assignments[day] = shift;
    emp.assigned_days++;
}

void schedule_day(const string& day, vector<Employee>& employees, Schedule& schedule) {
    vector<Employee*> available;
    for (auto& emp : employees) {
        if (can_work(emp, day)) {
            available.push_back(&emp);
        }
    }

    sort(available.begin(), available.end(), [](Employee* a, Employee* b) {
        if (a->assigned_days != b->assigned_days) return a->assigned_days < b->assigned_days;
        return a->name < b->name;
    });

    for (auto* emp : available) {
        for (const auto& pref : emp->preferences.at(day)) {
            if (schedule[day][pref].size() < CAPACITY_PER_SHIFT) {
                assign_shift(*emp, day, pref, schedule);
                break;
            }
        }
    }

    for (const string& shift : SHIFTS) {
        while (schedule[day][shift].size() < MIN_PER_SHIFT) {
            Employee* candidate = nullptr;
            for (auto& emp : employees) {
                if (!can_work(emp, day)) continue;
                if (find(schedule[day][shift].begin(), schedule[day][shift].end(), emp.name) != schedule[day][shift].end()) continue;
                int rank = find(emp.preferences.at(day).begin(), emp.preferences.at(day).end(), shift) - emp.preferences.at(day).begin();
                if (!candidate || rank < find(candidate->preferences.at(day).begin(), candidate->preferences.at(day).end(), shift) - candidate->preferences.at(day).begin() || (rank == find(candidate->preferences.at(day).begin(), candidate->preferences.at(day).end(), shift) - candidate->preferences.at(day).begin() && emp.assigned_days < candidate->assigned_days)) {
                    candidate = &emp;
                }
            }
            if (!candidate) break;
            assign_shift(*candidate, day, shift, schedule);
        }
    }
}

Schedule schedule_week(vector<Employee>& employees) {
    Schedule schedule = initialize_schedule();
    for (const string& day : DAYS) {
        schedule_day(day, employees, schedule);
    }
    return schedule;
}

void print_schedule(const Schedule& schedule) {
    cout << "\nFinal Weekly Schedule:\n";
    for (const string& day : DAYS) {
        cout << day << ":\n";
        for (const string& shift : SHIFTS) {
            cout << "  " << shift << ": ";
            const auto& names = schedule.at(day).at(shift);
            if (!names.empty()) {
                for (size_t i = 0; i < names.size(); i++) {
                    cout << names[i];
                    if (i + 1 < names.size()) cout << ", ";
                }
                cout << "\n";
            } else {
                cout << "(no assignment)\n";
            }
        }
        cout << "\n";
    }
}

void print_summary(const vector<Employee>& employees) {
    cout << "Employee assignment summary:\n";
    for (const auto& emp : employees) {
        cout << "  " << emp.name << ": " << emp.assigned_days << " days\n";
    }
    cout << "\n";
}

int main() {
    cout << "Employee Schedule Manager" << endl;
    vector<Employee> employees = demo_employees();
    Schedule schedule = schedule_week(employees);
    print_schedule(schedule);
    print_summary(employees);
    return 0;
}
